File: frontend/src/types/Geral/geral.ts
// ============================================================================
// TIPOS DO CADASTRO GERAL (Clientes, Fornecedores, Transportadoras, Vendedores)
// ============================================================================

/**
 * DTO para listagem simplificada
 */
export interface GeralListDto {
  sequenciaDoGeral: number;
  razaoSocial: string;
  nomeFantasia: string;
  cpfECnpj: string;
  cidade: string;
  uf: string;
  fone1: string;
  email: string;
  cliente: boolean;
  fornecedor: boolean;
  transportadora: boolean;
  vendedor: boolean;
  inativo: boolean;
  tipo: number; // 0=PF, 1=PJ
}

/**
 * DTO para criação
 */
export interface GeralCreateDto {
  // Tipos (flags)
  cliente: boolean;
  fornecedor: boolean;
  despesa: boolean;
  imposto: boolean;
  transportadora: boolean;
  vendedor: boolean;

  // Dados Principais
  razaoSocial: string;
  nomeFantasia: string;
  tipo: number; // 0=PF, 1=PJ
  cpfECnpj: string;
  rgEIe: string;
  codigoDoSuframa: string;
  codigoDaAntt: string;

  // Endereço
  endereco: string;
  numeroDoEndereco: string;
  complemento: string;
  bairro: string;
  caixaPostal: string;
  sequenciaDoMunicipio: number;
  cep: string;
  sequenciaDoPais: number;

  // Contato
  fone1: string;
  fone2: string;
  fax: string;
  celular: string;
  contato: string;
  email: string;
  homePage: string;

  // Observações
  observacao: string;

  // Flags fiscais
  revenda: boolean;
  isento: boolean;
  orgonPublico: boolean;
  empresaProdutor: boolean;
  cumulativo: boolean;
  inativo: boolean;

  // Vendedor associado
  sequenciaDoVendedor: number;
  intermediarioDoVendedor: string;

  // Dados de Cobrança (aba 2)
  enderecoDeCobranca: string;
  numeroDoEnderecoDeCobranca: string;
  complementoDaCobranca: string;
  bairroDeCobranca: string;
  caixaPostalDaCobranca: string;
  sequenciaMunicipioCobranca: number;
  cepDeCobranca: string;

  // Dados Bancários (aba 2)
  nomeDoBanco1: string;
  agenciaDoBanco1: string;
  contaCorrenteDoBanco1: string;
  nomeDoCorrentistaDoBanco1: string;
  nomeDoBanco2: string;
  agenciaDoBanco2: string;
  contaCorrenteDoBanco2: string;
  nomeDoCorrentistaDoBanco2: string;

  // Dados adicionais
  dataDeNascimento?: string;
  codigoContabil: number;
  codigoAdiantamento: number;
  salBruto: number;
  whatsAppSincronizado: boolean;
}

/**
 * DTO para atualização
 */
export interface GeralUpdateDto extends GeralCreateDto {
  sequenciaDoGeral: number;
}

/**
 * DTO detalhado (resposta da API)
 */
export interface GeralDetailDto extends GeralUpdateDto {
  dataDoCadastro?: string;
  usuDaAlteracao: string;
  municipioNome: string;
  municipioUf: string;
  municipioCobrancaNome: string;
  municipioCobrancaUf: string;
  vendedorNome: string;
}

// PagedResult importado de Common
export type { PagedResult } from '../Common/common';

/**
 * Filtros para listagem
 */
export interface GeralFiltros {
  pageNumber?: number;
  pageSize?: number;
  busca?: string;
  cliente?: boolean;
  fornecedor?: boolean;
  transportadora?: boolean;
  vendedor?: boolean;
  incluirInativos?: boolean;
}

/**
 * Tipos de entidade para navegação
 */
export type TipoEntidade = 'todos' | 'cliente' | 'fornecedor' | 'transportadora' | 'vendedor';

/**
 * Labels para os tipos
 */
export const TIPO_LABELS: Record<TipoEntidade, string> = {
  todos: 'Todos',
  cliente: 'Clientes',
  fornecedor: 'Fornecedores',
  transportadora: 'Transportadoras',
  vendedor: 'Vendedores',
};

/**
 * Cores para badges de tipo
 */
export const TIPO_CORES: Record<string, string> = {
  cliente: 'bg-blue-100 text-blue-800',
  fornecedor: 'bg-green-100 text-green-800',
  transportadora: 'bg-purple-100 text-purple-800',
  vendedor: 'bg-orange-100 text-orange-800',
  despesa: 'bg-red-100 text-red-800',
  imposto: 'bg-yellow-100 text-yellow-800',
};

/**
 * Siglas curtas para badges compactos
 */
export const TIPO_SIGLAS: Record<string, string> = {
  cliente: 'CLI',
  fornecedor: 'FOR',
  transportadora: 'TRANSP',
  vendedor: 'VEN',
  despesa: 'DES',
  imposto: 'IMP',
};

/**
 * Retorna os tipos ativos de um registro
 */
export function getTiposAtivos(item: GeralListDto): string[] {
  const tipos: string[] = [];
  if (item.cliente) tipos.push('cliente');
  if (item.fornecedor) tipos.push('fornecedor');
  if (item.transportadora) tipos.push('transportadora');
  if (item.vendedor) tipos.push('vendedor');
  return tipos;
}

/**
 * Formata o tipo pessoa
 */
export function formatTipoPessoa(tipo: number): string {
  return tipo === 0 ? 'Pessoa Física' : 'Pessoa Jurídica';
}

/**
 * Formata o label CPF/CNPJ
 */
export function getLabelDocumento(tipo: number): string {
  return tipo === 0 ? 'CPF' : 'CNPJ';
}

/**
 * Formata o label RG/IE
 */
export function getLabelIdentidade(tipo: number): string {
  return tipo === 0 ? 'RG' : 'I.E.';
}

/**
 * Valores padrão para novo registro
 */
export const GERAL_DEFAULT: GeralCreateDto = {
  cliente: false,
  fornecedor: false,
  despesa: false,
  imposto: false,
  transportadora: false,
  vendedor: false,
  razaoSocial: '',
  nomeFantasia: '',
  tipo: 1, // Jurídica por padrão
  cpfECnpj: '',
  rgEIe: '',
  codigoDoSuframa: '',
  codigoDaAntt: '',
  endereco: '',
  numeroDoEndereco: '',
  complemento: '',
  bairro: '',
  caixaPostal: '',
  sequenciaDoMunicipio: 0,
  cep: '',
  sequenciaDoPais: 1,
  fone1: '',
  fone2: '',
  fax: '',
  celular: '',
  contato: '',
  email: '',
  homePage: '',
  observacao: '',
  revenda: false,
  isento: false,
  orgonPublico: false,
  empresaProdutor: false,
  cumulativo: false,
  inativo: false,
  sequenciaDoVendedor: 0,
  intermediarioDoVendedor: '',
  enderecoDeCobranca: '',
  numeroDoEnderecoDeCobranca: '',
  complementoDaCobranca: '',
  bairroDeCobranca: '',
  caixaPostalDaCobranca: '',
  sequenciaMunicipioCobranca: 0,
  cepDeCobranca: '',
  nomeDoBanco1: '',
  agenciaDoBanco1: '',
  contaCorrenteDoBanco1: '',
  nomeDoCorrentistaDoBanco1: '',
  nomeDoBanco2: '',
  agenciaDoBanco2: '',
  contaCorrenteDoBanco2: '',
  nomeDoCorrentistaDoBanco2: '',
  codigoContabil: 0,
  codigoAdiantamento: 0,
  salBruto: 0,
  whatsAppSincronizado: false,
};

File: SistemaEmpresas/Models/Geral.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace SistemaEmpresas.Models;

[Table("Geral")]
public partial class Geral
{
    public bool Cliente { get; set; }

    public bool Fornecedor { get; set; }

    public bool Despesa { get; set; }

    public bool Imposto { get; set; }

    public bool Transportadora { get; set; }

    public bool Vendedor { get; set; }

    [Key]
    [Column("Seqüência do Geral")]
    public int SequenciaDoGeral { get; set; }

    [Column("Razão Social")]
    [StringLength(60)]
    [Unicode(false)]
    public string RazaoSocial { get; set; } = null!;

    [Column("Nome Fantasia")]
    [StringLength(60)]
    [Unicode(false)]
    public string NomeFantasia { get; set; } = null!;

    public short Tipo { get; set; }

    [Column("Endereço")]
    [StringLength(100)]
    [Unicode(false)]
    public string Endereco { get; set; } = null!;

    [StringLength(100)]
    [Unicode(false)]
    public string Complemento { get; set; } = null!;

    [Column("Número do Endereço")]
    [StringLength(10)]
    [Unicode(false)]
    public string NumeroDoEndereco { get; set; } = null!;

    [StringLength(50)]
    [Unicode(false)]
    public string Bairro { get; set; } = null!;

    [Column("Caixa Postal")]
    [StringLength(30)]
    [Unicode(false)]
    public string CaixaPostal { get; set; } = null!;

    [Column("Seqüência do Município")]
    public int SequenciaDoMunicipio { get; set; }

    [Column("CEP")]
    [StringLength(9)]
    [Unicode(false)]
    public string Cep { get; set; } = null!;

    [Column("Fone 1")]
    [StringLength(14)]
    [Unicode(false)]
    public string? Fone1 { get; set; }

    [Column("Fone 2")]
    [StringLength(14)]
    [Unicode(false)]
    public string? Fone2 { get; set; }

    [StringLength(14)]
    [Unicode(false)]
    public string? Fax { get; set; }

    [StringLength(14)]
    [Unicode(false)]
    public string? Celular { get; set; }

    [StringLength(45)]
    [Unicode(false)]
    public string Contato { get; set; } = null!;

    [StringLength(255)]
    [Unicode(false)]
    public string Email { get; set; } = null!;

    [Column("Home Page")]
    [StringLength(60)]
    [Unicode(false)]
    public string HomePage { get; set; } = null!;

    [Column("Código do Suframa")]
    [StringLength(9)]
    [Unicode(false)]
    public string CodigoDoSuframa { get; set; } = null!;

    [Column("Código da ANTT")]
    [StringLength(20)]
    [Unicode(false)]
    public string CodigoDaAntt { get; set; } = null!;

    [Column("CPF e CNPJ")]
    [StringLength(20)]
    [Unicode(false)]
    public string CpfECnpj { get; set; } = null!;

    [Column("RG e IE")]
    [StringLength(20)]
    [Unicode(false)]
    public string RgEIe { get; set; } = null!;

    [Column("Observação", TypeName = "text")]
    public string Observacao { get; set; } = null!;

    [Column("Endereço de Cobrança")]
    [StringLength(62)]
    [Unicode(false)]
    public string EnderecoDeCobranca { get; set; } = null!;

    [Column("Número do Endereço de Cobrança")]
    [StringLength(20)]
    [Unicode(false)]
    public string NumeroDoEnderecoDeCobranca { get; set; } = null!;

    [Column("Seqüência Município Cobrança")]
    public int SequenciaMunicipioCobranca { get; set; }

    [Column("Cep de Cobrança")]
    [StringLength(9)]
    [Unicode(false)]
    public string CepDeCobranca { get; set; } = null!;

    [Column("Bairro de Cobrança")]
    [StringLength(50)]
    [Unicode(false)]
    public string BairroDeCobranca { get; set; } = null!;

    [Column("Complemento da Cobrança")]
    [StringLength(30)]
    [Unicode(false)]
    public string ComplementoDaCobranca { get; set; } = null!;

    [Column("Caixa Postal da Cobrança")]
    [StringLength(30)]
    [Unicode(false)]
    public string CaixaPostalDaCobranca { get; set; } = null!;

    [Column("Seqüência do Vendedor")]
    public int SequenciaDoVendedor { get; set; }

    [Column("Intermediário do Vendedor")]
    [StringLength(40)]
    [Unicode(false)]
    public string IntermediarioDoVendedor { get; set; } = null!;

    [Column("Nome do Banco 1")]
    [StringLength(20)]
    [Unicode(false)]
    public string NomeDoBanco1 { get; set; } = null!;

    [Column("Nome do Banco 2")]
    [StringLength(20)]
    [Unicode(false)]
    public string NomeDoBanco2 { get; set; } = null!;

    [Column("Agência do Banco 1")]
    [StringLength(20)]
    [Unicode(false)]
    public string AgenciaDoBanco1 { get; set; } = null!;

    [Column("Agência do Banco 2")]
    [StringLength(20)]
    [Unicode(false)]
    public string AgenciaDoBanco2 { get; set; } = null!;

    [Column("Conta Corrente do Banco 1")]
    [StringLength(15)]
    [Unicode(false)]
    public string ContaCorrenteDoBanco1 { get; set; } = null!;

    [Column("Conta Corrente do Banco 2")]
    [StringLength(15)]
    [Unicode(false)]
    public string ContaCorrenteDoBanco2 { get; set; } = null!;

    [Column("Nome do Correntista do Banco 1")]
    [StringLength(60)]
    [Unicode(false)]
    public string NomeDoCorrentistaDoBanco1 { get; set; } = null!;

    [Column("Nome do Correntista do Banco 2")]
    [StringLength(60)]
    [Unicode(false)]
    public string NomeDoCorrentistaDoBanco2 { get; set; } = null!;

    public bool Inativo { get; set; }

    public bool Revenda { get; set; }

    public bool Isento { get; set; }

    [Column("Data do Cadastro", TypeName = "datetime")]
    public DateTime? DataDoCadastro { get; set; }

    [Column("Seqüência do País")]
    public int SequenciaDoPais { get; set; }

    [Column("Orgõn Publico")]
    public bool OrgonPublico { get; set; }

    public bool Cumulativo { get; set; }

    [Column("Empresa Produtor")]
    public bool EmpresaProdutor { get; set; }

    [Column("Usu da Alteração")]
    [StringLength(40)]
    [Unicode(false)]
    public string UsuDaAlteracao { get; set; } = null!;

    [Column("Data de Nascimento", TypeName = "datetime")]
    public DateTime? DataDeNascimento { get; set; }

    [Column("Codigo Contabil")]
    public int CodigoContabil { get; set; }

    [Column("Codigo Adiantamento")]
    public int CodigoAdiantamento { get; set; }

    [Column("Sal bruto", TypeName = "decimal(10, 2)")]
    public decimal SalBruto { get; set; }

    [Column("Importou no Zap")]
    public bool WhatsAppSincronizado { get; set; }

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<AdicaoDaDeclaracao> AdicoesDaDeclaracaos { get; set; } = new List<AdicaoDaDeclaracao>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<BaixaDoEstoqueContabil> BaixaDoEstoqueContabils { get; set; } = new List<BaixaDoEstoqueContabil>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<DeclaracaoDeImportacao> DeclaracoesDeImportacaos { get; set; } = new List<DeclaracaoDeImportacao>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<EntradaConta> EntradaConta { get; set; } = new List<EntradaConta>();

    [InverseProperty("SequenciaDoVendedorNavigation")]
    public virtual ICollection<Geral> InverseSequenciaDoVendedorNavigation { get; set; } = new List<Geral>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<ManutencaoConta> ManutencaoConta { get; set; } = new List<ManutencaoConta>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<MovimentoDoEstoqueContabil> MovimentoDoEstoqueContabils { get; set; } = new List<MovimentoDoEstoqueContabil>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<MovimentoDoEstoque> MovimentoDoEstoques { get; set; } = new List<MovimentoDoEstoque>();

    [InverseProperty("SequenciaDaTransportadoraNavigation")]
    public virtual ICollection<NotaFiscal> NotaFiscalSequenciaDaTransportadoraNavigations { get; set; } = new List<NotaFiscal>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<NotaFiscal> NotaFiscalSequenciaDoGeralNavigations { get; set; } = new List<NotaFiscal>();

    [InverseProperty("SequenciaDoVendedorNavigation")]
    public virtual ICollection<NotaFiscal> NotaFiscalSequenciaDoVendedorNavigations { get; set; } = new List<NotaFiscal>();

    [InverseProperty("SequenciaDaTransportadoraNavigation")]
    public virtual ICollection<NovaLicitacao> NovaLicitacaoSequenciaDaTransportadoraNavigations { get; set; } = new List<NovaLicitacao>();

    [InverseProperty("SequenciaDoFornecedorNavigation")]
    public virtual ICollection<NovaLicitacao> NovaLicitacaoSequenciaDoFornecedorNavigations { get; set; } = new List<NovaLicitacao>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<OrdemDeServico> OrdemDeServicoSequenciaDoGeralNavigations { get; set; } = new List<OrdemDeServico>();

    [InverseProperty("SequenciaDoVendedorNavigation")]
    public virtual ICollection<OrdemDeServico> OrdemDeServicoSequenciaDoVendedorNavigations { get; set; } = new List<OrdemDeServico>();

    [InverseProperty("SequenciaDaTransportadoraNavigation")]
    public virtual ICollection<Orcamento> OrcamentoSequenciaDaTransportadoraNavigations { get; set; } = new List<Orcamento>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<Orcamento> OrcamentoSequenciaDoGeralNavigations { get; set; } = new List<Orcamento>();

    [InverseProperty("SequenciaDoVendedorNavigation")]
    public virtual ICollection<Orcamento> OrcamentoSequenciaDoVendedorNavigations { get; set; } = new List<Orcamento>();

    [InverseProperty("SequenciaDaTransportadoraNavigation")]
    public virtual ICollection<Pedido> PedidoSequenciaDaTransportadoraNavigations { get; set; } = new List<Pedido>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<Pedido> PedidoSequenciaDoGeralNavigations { get; set; } = new List<Pedido>();

    [InverseProperty("SequenciaDoVendedorNavigation")]
    public virtual ICollection<Pedido> PedidoSequenciaDoVendedorNavigations { get; set; } = new List<Pedido>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<PropriedadeDoGeral> PropriedadesDoGerals { get; set; } = new List<PropriedadeDoGeral>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<RelatorioDeViagem> RelatorioDeViagems { get; set; } = new List<RelatorioDeViagem>();

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<Requisicao> Requisicaos { get; set; } = new List<Requisicao>();

    [ForeignKey("SequenciaDoMunicipio")]
    [InverseProperty("GeralSequenciaDoMunicipioNavigations")]
    public virtual Municipio SequenciaDoMunicipioNavigation { get; set; } = null!;

    [ForeignKey("SequenciaDoPais")]
    [InverseProperty("Gerals")]
    public virtual Pais SequenciaDoPaisNavigation { get; set; } = null!;

    [ForeignKey("SequenciaDoVendedor")]
    [InverseProperty("InverseSequenciaDoVendedorNavigation")]
    public virtual Geral SequenciaDoVendedorNavigation { get; set; } = null!;

    [ForeignKey("SequenciaMunicipioCobranca")]
    [InverseProperty("GeralSequenciaMunicipioCobrancaNavigations")]
    public virtual Municipio SequenciaMunicipioCobrancaNavigation { get; set; } = null!;

    [InverseProperty("SequenciaDoGeralNavigation")]
    public virtual ICollection<VinculaPedidoOrcamento> VinculaPedidoOrcamentos { get; set; } = new List<VinculaPedidoOrcamento>();
}
